### PR TITLE
Move R tests to GitHub Actions

### LINKED ISSUE #3047: [CI] move R tests from Travis to github actions
As discussed with Javier, this change should help unblock https://github.com/mlflow/mlflow/pull/2367

### PR DESCRIPTION
**Pull request by lorenzwalthert**
## What changes are proposed in this pull request?

Move the R tests from Travis to GitHub actions. This was previously attempted unsuccessfully because a missing `libcurl` system requirement was not installed in https://github.com/mlflow/mlflow/pull/2724, so I am basically completing some of @aarondav's work here. This closes #3047. Also, it is supposed to unblock #2367 because of newer gcc versions.

## How is this patch tested?

Github Actions 😊. I mainly copied the scripts, so the same should be tested as we tested on Travis. I managed to cache R package installation as described in https://github.com/r-lib/actions/tree/master/examples and code coverage also seems to work.

Some things up for debate as far as the CI specification goes are:

- R versions (currently only latest).
- platform: currently only ubuntu.
- platform version: currently only 18.04. 16.04 did also work (see 0176c59).

I am unsure how restricted you are in terms of build time. Please let me know if I should add more builds to the matrix.

TODO before merge

- Re-add all non-R tests. I disabled them for testing the R build. 

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [x] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by smurching** [comment]
@lorenzwalthert this is fantastic! Thanks for the contribution, and let us know when it's ready for review :)

**Comment by smurching** [comment]
BTW, I think testing the latest R version on Ubuntu 18.04 as you've done should suffice for now - I don't think we use any platform-specific R APIs (or unstable R APIs) that would introduce significant risk if we don't test against other OSes or R versions

**Comment by lorenzwalthert** [comment]
Ok, thanks @smurching. I think this is ready for merge. For some reason, the lintr action for python now does not pass anymore, but it's unrelated to what I did, so I hope we can merge this soon. I might do some more work on the R CI/CD to make it a bit cleaner after #2367.

**Comment by lorenzwalthert** [comment]
@smurching all lights green after merging master into this branch. I think this is ready to merge from my side.

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -27,7 +27,50 @@ jobs:
...
+    - name: Query dependencies
+      run: |
+        install.packages('remotes')
+        saveRDS(remotes::dev_package_deps("mlflow/R/mlflow", dependencies = TRUE), ".github/depends.Rds", version = 2)
+        writeLines(sprintf("R-%i.%i", getRversion()$major, getRversion()$minor), ".github/R-version")
+      shell: Rscript {0}
 
+    - name: Cache R packages
+      if: runner.os != 'Windows'
+      uses: actions/cache@v1
```
Comment (smurching) [inline .github/workflows/master.yml:44]: Looks like the latest version of the `cache` action is v2: https://github.com/actions/cache#whats-new

Out of curiosity, any reason we're using v1 here? If not, maybe we could just upgrade to v2 preemptively

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -27,7 +27,50 @@ jobs:
         export PATH="$HOME/miniconda/bin:$PATH"
         source activate test-environment
         ./lint.sh
+  r:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@master
+    - uses: r-lib/actions/setup-r@v1
+    - name: Query dependencies
```
Comment (smurching) [inline .github/workflows/master.yml:35]:
```suggestion
    # This step dumps the current set of R dependencies and R version into files to be used
    # as a cache key when caching/restoring R dependencies.
    - name: Query dependencies
```

Would be good to add a clarifying comment explaining the purpose of this step

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -27,7 +27,50 @@ jobs:
...
+        saveRDS(remotes::dev_package_deps("mlflow/R/mlflow", dependencies = TRUE), ".github/depends.Rds", version = 2)
+        writeLines(sprintf("R-%i.%i", getRversion()$major, getRversion()$minor), ".github/R-version")
+      shell: Rscript {0}
 
+    - name: Cache R packages
+      if: runner.os != 'Windows'
+      uses: actions/cache@v1
+      with:
+        path: ${{ env.R_LIBS_USER }}
+        key: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-${{ hashFiles('.github/depends.Rds') }}
```
Comment (smurching) [inline .github/workflows/master.yml:47]:
```suggestion
        # We cache R dependencies based on a tuple of the current OS, the R version, and the list of
        # R dependencies 
        key: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-${{ hashFiles('.github/depends.Rds') }}
```

**Review by smurching — COMMENTED** [review]
Some small comments, otherwise largely LGTM - thanks @lorenzwalthert!

File: .github/workflows/master.yml (hunk 1 above)
Comment (lorenzwalthert) [reply .github/workflows/master.yml:44]: right. TBH, I just copied from the template from r-lib/actions. Let's try `v2`.

File: .github/workflows/master.yml (hunk 2 above)
Comment (lorenzwalthert) [reply .github/workflows/master.yml:35]: Sure. Also, we might want to use the `r-cache` action (https://github.com/r-lib/actions/issues/114) once it's implemented.

**Comment by lorenzwalthert** [comment]
Thanks @smurching, I adapted to your suggestions.

File: .github/workflows/master.yml (hunk 2 above)
Comment (smurching) [reply .github/workflows/master.yml:37]: Yep makes sense!

**Review by smurching — APPROVED** [review]
LGTM

**Comment by smurching** [comment]
Thanks again @lorenzwalthert!

**Comment by javierluraschi** [comment]
Thank you, @lorenzwalthert  and @smurching!

**Pull request merged**